qiskit_experiments/database_service/db_experiment_data.py

```diff
@@ -1055,52 +1074,80 @@ def cancel_jobs(self, ids: Optional[List[str]] = None) -> bool:
+
+            # Get IDs of futures to cancel
             if ids is None:
-                jobs = self._jobs.values()
-            else:
-                jobs = [self._jobs[jid] for jid in ids]
+                ids = self._jobs.keys()
+            elif isinstance(ids, str):
+                ids = [ids]
+            jobs = [self._jobs[jid] for jid in ids]
 
-            for job in jobs:
+            # Attempt to cancel callbacks
+            for jid, job in zip(ids, jobs):
                 if job and job.status() not in JOB_FINAL_STATES:
                     try:
                         job.cancel()
-                        LOG.debug("Cancelled job [Job ID: %s]", job.job_id())
+                        LOG.warning("Cancelled job [Job ID: %s]", jid)
                     except Exception as err:  # pylint: disable=broad-except
                         cancelled = False
-                        LOG.warning("Unable to cancel job [Job ID: %s]:\n%s", job.job_id(), err)
+                        LOG.warning("Unable to cancel job [Job ID: %s]:\n%s", jid, err)
+                        continue
+
+                # Remove done or cancelled job futures
+                if jid in self._job_futures:
+                    del self._job_futures[jid]
+
         return cancelled
 
-    def cancel_analysis(self, ids: Optional[List[str]] = None) -> bool:
+    def cancel_analysis(self, ids: Optional[Union[str, List[str]]] = None) -> bool:
         """Cancel any queued analysis callbacks.
 
         .. note::
             A currently running analysis callback cannot be cancelled.
 
         Args:
-            ids: Analysis callbacks to cancel. If None all non-finished
+            ids: Analysis callback(s) to cancel. If None all non-finished
                  analysis will be cancelled.
 
         Returns:
             True if the specified analysis callbacks were successfully
             cancelled otherwise false.
         """
-        cancelled = True
+        all_cancelled = True
+        # Lock analysis futures so we can't add more while trying to cancel
         with self._analysis_futures.lock:
+
+            # Get IDs of futures to cancel
             if ids is None:
-                ids = self._analysis_futures.keys()
-                futs = self._analysis_futures.values()
-            else:
-                futs = [self._analysis_futures[i] for i in ids]
-            for cid, fut in zip(ids, futs):
-                if fut.done():
-                    continue
-                if fut.cancel():
-                    self._analysis_callbacks[cid].status = AnalysisStatus.CANCELLED
-                    LOG.info("Cancelled analysis callback [Analysis ID: %s]", cid)
+                # We reverse iterate since later callbacks are queued pending earlier ones
+                # and can have their future directly cancelled.
+                ids = list(reversed(self._analysis_futures.keys()))
+            elif isinstance(ids, str):
+                ids = [ids]
```

Comment:
The same mechanism should be applied to this. I prefer 
```python
for cid, callback in reversed(self._analysis_callbacks.items()):
    if ids and cid not in ids:
        continue
    ...
```
same for job cancel logic (this one is just my preference)